code/trans.c

```diff
@@ -0,0 +1,377 @@
+        *refIO = refNew;
+      }
+    }
+  }
+
+  /* Now progress to a normal GC fix. */
+  /* TODO: Make a clean interface to this kind of dynamic binding. */
+  ss->fix = ss->arena->emergency ? SegFixEmergency : SegFix;
+  TRACE_SCAN_BEGIN(ss) {
+    res = TRACE_FIX12(ss, refIO);
+  } TRACE_SCAN_END(ss);
+  ss->fix = transformFix;
+
+  return res;
+}
+
+
+static void transformCondemn(void *closure, Word old, void *value)
+{
+  Seg seg;
+  GenDesc gen;
+  Bool b;
+  Trace trace = closure;
+
+  AVERT(Trace, trace);
+  UNUSED(value);
+
+  /* Find segment containing old address. */
+  b = SegOfAddr(&seg, trace->arena, (Ref)old);
+  AVER(b); /* old refs must be in managed memory, else client param error */
+
+  /* Condemn generation containing seg if not already condemned. */
+  gen = PoolSegPoolGen(SegPool(seg), seg)->gen;
+  AVERT(GenDesc, gen);
+  if (RingIsSingle(&gen->trace[trace->ti].traceRing))
+    GenDescStartTrace(gen, trace);
+}
+
+
+Res TransformApply(Bool *appliedReturn, Transform transform)
+{
+  Res res;
+  Arena arena;
+  Globals globals;
+  Trace trace;
+  double mortality;
+
+  AVER(appliedReturn != NULL);
+  AVERT(Transform, transform);
+
+  arena = TransformArena(transform);
+
+  /* If there have been any flips since the transform was created, the old
+     and new pointers will be invalid, since they are not scanned as roots.
+     The client program must park the arena before applying the transform. */
+  if (transform->epoch != ArenaEpoch(arena))
+    return ResPARAM;
+
+  globals = ArenaGlobals(arena);
+  AVERT(Globals, globals);
+
+  ArenaPark(globals);
+  
+  res = TraceCreate(&trace, arena, TraceStartWhyEXTENSION);
+  AVER(res == ResOK); /* parking should make a trace available */
+  if (res != ResOK)
+    return res;
+
+  /* Condemn the generations containing the transform's old objects,
+     so that all references to them are scanned. */
+  TraceCondemnStart(trace);
+  TableMap(transform->oldToNew, transformCondemn, trace);
+  res = TraceCondemnEnd(&mortality, trace);
+  if (res != ResOK) {
+    /* Nothing to transform. */
+    TraceDestroyInit(trace);
+    goto done;
+  }
+
+  trace->fix = transformFix;
+  trace->fixClosure = transform;
+
+  res = TraceStart(trace, 1.0, 0.0);
+  AVER(res == ResOK); /* transformFix can't fail */
+  
+  /* If transformFix during traceFlip found ambiguous references and
+     aborted the transform then the rest of the trace is just a normal GC. 
+     Note that aborting a trace part-way through is pretty much impossible
+     without corrupting the mutator graph.  We could safely
+         if (transform->aborted) {
+           trace->fix = PoolFix;
+           trace->fixClosure = NULL;
+         }
+   */
+  
+  /* Force the trace to complete now. */
+  ArenaPark(globals);
+
+done:
+  if (transform->aborted) {
+    *appliedReturn = FALSE;
+  } else {
+    *appliedReturn = TRUE;
+    /* I'm not sure why the interface is defined this way. RB 2012-08-03 */
+    TransformDestroy(transform);
```

Comment:
This is an annoying wart on the interface — I think we should fix this before making the interface public, even though this requires the customer to update their code.